### ISSUE TITLE
Alternating between variables

### ISSUE THREAD

**Issue by matillda123** [patrick-kidger/optimistix#141]
Hey,

is it possible to let optimizers alternate between the optimization of different variables? 
I have two sets of very "different" variables whose combined optimization seems problematic. Thus i was wondering if it may be an improvement to let the optimizer ignore them in an alternating fashion. (e.g. optimizing for A every even iteration and for B every odd iteration)

**Comment by matillda123** [comment]
@johannahaffner ah yes that seems sensible. So you suggest following Approach 2 in [Docs Custom Solver](https://docs.kidger.site/optimistix/examples/custom_solver/) where one defines a whole new query method and so on?

I realized that certainly some solvers reuse the information from the state-object. Letting solvers make alternating steps will cause the information they use to be outdated, right? So after every solver has made one step one has to update some information in all state objects.

**Comment by johannahaffner** [comment]
So solvers write the function information for the last accepted step into their state. This could, for example, include the Hessian, gradient and function value at the last accepted step (or any one of those three). 

When alternating between two optimisation variables, then the function information in the states would indeed be "one step behind" - this would affect differences in $f$ computed in `solver.step`. 
(The previous function value considered in solver A would not include the update to the argument computer by solver B, but the current function value would.)

If it is important to you to try to synchronise that, then you might want to do tree-surgery on the function info object in the states - I think this would be cleaner than the merging approach, but it means that your two solvers have to use function information of the same type.

I don't think that this requires overriding or implementing a custom query, since you are interested in using solvers, rather directly working with descents.

**Comment by jpbrodrick89** [comment]
Just chiming in and wondering whether SGD might be a good method here as that doesn't typically require info on the Hessian and instead treats the gradient obtained as "approximate"?

**Comment by jpbrodrick89** [comment]
Also maybe add some kind of constraints that the step taken can't make the other variable worse?

**Comment by johannahaffner** [comment]
> Also maybe add some kind of constraints that the step taken can't make the other variable worse?

We only know if the other step got worse after calling the other (inner) solver again - definitely doable, but no longer just a wrapper: we'd have to add a layer of accept/reject criteria at the outer solver level.

**Comment by matillda123** [comment]
This is what I've come up with now. 

```
from collections.abc import Callable
from optimistix._solver.gradient_methods import _GradientDescentState
import optimistix as optx
import equinox as eqx



class AlternatingSolverState(eqx.Module):
    state_1: _GradientDescentState
    state_2: _GradientDescentState


class AlternatingSolver(optx.AbstractMinimiser):
    solver: optx.GradientDescent

    rtol: float=solver.rtol
    atol: float=solver.atol
    norm: Callable=solver.norm

    def init(self, fn, y, args, options, f_struct, aux_struct, tags):
        y_1, y_2 = y
        args_1, args_2 = y_2, y_1
        state_1=self.solver.init(fn, y_1, args_1, options, f_struct, aux_struct, tags)
        state_2=self.solver.init(fn, y_2, args_2, options, f_struct, aux_struct, tags)
        return AlternatingSolverState(state_1, state_2)
    

    def step(self, fn, y, args, options, state, tags):
        y_1, y_2 = y
        args_1, args_2 = y_2, y_1

        new_y_1, new_state_1, new_aux_1 = self.solver.step(fn, y_1, args_1, options, state.state_1, tags)
        new_y_2, new_state_2, new_aux_2 = self.solver.step(fn, y_2, args_2, options, state.state_2, tags)
        return (new_y_1, new_y_2), AlternatingSolverState(new_state_1, new_state_2), (new_aux_1, new_aux_2)
    

    def terminate(self, fn, y, args, options, state, tags):
        y_1, y_2 = y
        args_1, args_2 = y_2, y_1

        state_1_terminate, state_1_result = self.solver.terminate(fn, y_1, args_1, options, state.state_1, tags)
        state_2_terminate, state_2_result = self.solver.terminate(fn, y_2, args_2, options, state.state_2, tags)
        return jnp.logical_and(state_1_terminate, state_2_terminate), state_2_result


    def postprocess(self, fn, y, aux, args, options, state, tags, result):
        return y, aux, {}
```



```
import jax.numpy as jnp
import numpy as np

y_init=jnp.array(np.random.uniform(-10, 10, size=(2,5,5)))

def test_func(x, y):
    return jnp.sum((x-1)**2+(y+1)**2)


test=AlternatingSolver(optx.GradientDescent(learning_rate=0.1, rtol=1e-6, atol=1e-6))
solution=optx.minimise(test_func, test, y0=y_init)
```

But I am getting the exact same behaviour as in the first failed example I posted. Namely that there is a TypeError when one does not specify something other than None for args and if one does there is a ValueError raised by equinox due to some shape inconsistencies.

The TypeError due to args=None seems to be more concerning to me since that means that the arguments args_1 and args_2 that I am defining in the step() method are not actually used??
The ValueError from equinox may be due to the fact that I am not manipulating f_struct in the init() method?

**Comment by matillda123** [comment]
I resorted to using alternating interactive solves, which seems to work. 

```
import jax
import jax.numpy as jnp
from equinox import Partial
import numpy as np


def test_func(x,y):
    return jnp.sum((x-1)**2+(y+1)**2), None

def test_func_x(x, y):
    return test_func(x, y)

def test_func_y(y, x):
    return test_func(x, y)



def fix_multiple_argument_error_from_partial(fn, y, args, state, options, tags, function):
    return function(fn, y, args, options, state, tags)
```

```
solver=optx.GradientDescent(learning_rate=0.1, rtol=1e-6, atol=1e-6)
y_init=jnp.array(np.random.uniform(-10,10,size=(2,5,5)))


options = dict()
f_struct = jax.ShapeDtypeStruct((), jnp.float32)
aux_struct = None
tags = frozenset()


state_1 = solver.init(fn=test_func_x, y=y_init[0], args=y_init[1], options=options, f_struct=f_struct, aux_struct=aux_struct, tags=tags)
state_2 = solver.init(fn=test_func_y, y=y_init[1], args=y_init[0], options=options, f_struct=f_struct, aux_struct=aux_struct, tags=tags)

step = Partial(fix_multiple_argument_error_from_partial, options=options, tags=tags, function=solver.step)


y_1, y_2 = y_init
for _ in range(100):
    y_1, state_1, aux_1 = step(test_func_x, y_1, y_2, state_1)
    y_2, state_2, aux_2 = step(test_func_y, y_2, y_1, state_2)
    
```

**Comment by johannahaffner** [comment]
Are you happy with that solution? It does seem to point to the issue being with compilation, or potentially with an early processing step (converting closures, casting the objective...): alternating positions of arguments interferes with the static shape requirement (this makes sense). There might be a clever way around this using custom filters if you have a strong preference for using one of our top-level APIs over an interactive solve. 
But I think the solution you came up with is probably the easiest to reason about! It is clear what gets optimised where.
